symforce/slam/imu_preintegration/imu_factor.h

```diff
@@ -0,0 +1,192 @@
+ *
+ * More generally, the gravity argument should yield the true acceleration when added to the
+ * accelerometer measurement (after the measurement has been converted to the world frame).
+ *
+ * Is a functor so as to store the preintegrated IMU measurements between two times.
+ * The residual computed is the local-coordinate difference between the final state (pose
+ * and velocity) and the state you would expect given the initial state and the IMU measurements.
+ *
+ * Based heavily on the on manifold ImuFactor found in GTSAM and on the paper:
+ *
+ * Christian Forster, Luca Carlone, Frank Dellaert, and Davide Scaramuzza,
+ * “IMU Preintegration on Manifold for Efficient Visual-Inertial Maximum-a-Posteriori Estimation”,
+ * Robotics: Science and Systems (RSS), 2015.
+ *
+ * Example Usage:
+ *
+ *   enum Var : char {
+ *     POSE = 'p',        // Pose3d
+ *     VELOCITY = 'v',    // Vector3d
+ *     ACCEL_BIAS = 'A',  // Vector3d
+ *     GYRO_BIAS = 'G',   // Vector3d
+ *     GRAVITY = 'g',     // Vector3d
+ *     EPSILON = 'e'      // Scalar
+ *   };
+ *
+ *   struct ImuMeasurement {
+ *     Eigen::Vector3d acceleration;
+ *     Eigen::Vector3d angular_velocity;
+ *     double duration;
+ *   };
+ *
+ *   int main() {
+ *     // Dummy function declarations
+ *     std::vector<ImuMeasurement> GetMeasurementsBetween(double start_time, double end_time);
+ *     std::vector<double> GetKeyFrameTimes();
+ *     Eigen::Vector3d GetAccelBiasEstimate(double time);
+ *     Eigen::Vector3d GetGyroBiasEstimate(double time);
+ *     void AppendOtherFactors(std::vector<sym::Factord> & factors);
+ *
+ *     // Example accelerometer and gyroscope covariances
+ *     const Eigen::Matrix3d accel_cov = 1e-5 * Eigen::Matrix3d::Identity();
+ *     const Eigen::Matrix3d gyro_cov = 1e-5 * Eigen::Matrix3d::Identity();
+ *
+ *     std::vector<sym::Factord> factors;
+ *
+ *     // GetKeyFrameTimes assumed to return at least 2 times
+ *     std::vector<double> key_frame_times = GetKeyFrameTimes();
+ *
+ *     // Integrating Imu measurements between keyframes, creating an ImuFactor for each
+ *     for (unsigned int i = 0; i < key_frame_times.size() - 1; i++) {
+ *       const double start_time = key_frame_times[i];
+ *       const double end_time = key_frame_times[i + 1];
+ *
+ *       const std::vector<ImuMeasurement> measurements = GetMeasurementsBetween(start_time,
+ *                                                                               end_time);
+ *
+ *       // ImuPreintegrator should be initialized with the best guesses for the IMU biases
+ *       sym::ImuPreintegrator<double> integrator(GetAccelBiasEstimate(start_time),
+ *                                                GetGyroBiasEstimate(start_time));
+ *       for (const ImuMeasurement& meas : measurements) {
+ *         // IntegrateMeasurement assumes IMU measures acceleration of magnitude |gravity| pointing
+ *         // away from the Earth when stationary
```

Comment:
Probably don't need this, since we explain the convention for the gravity vector above?

Reply:
I think I agree. And even then, it was slightly wrong, as `IntegrateMeasurement` doesn't actually need the `IMU` measurement to look like that. The fewer places we document, the fewer places we have to keep up to date.